src/services/identity/NotificationsService.ts

```diff
@@ -0,0 +1,249 @@
+import { FindOptions, ModelStatic, Op } from "sequelize"
+
+import { Notification, Site, Repo, SiteMember } from "@database/models"
+import {
+  NotificationType,
+  getNotificationExpiryDate,
+  getNotificationMessage,
+  getNotificationPriority,
+} from "@root/utils/notification-utils"
+
+const NUM_RECENT_NOTIFICATIONS = 6
+
+interface NotificationsServiceProps {
+  repository: ModelStatic<Notification>
+  siteMember: ModelStatic<SiteMember>
+}
+
+export interface NotificationResponse {
+  message: string
+  createdAt: Date
+  link: string
+  isRead: boolean
+  sourceUsername: string
+  type: string
+}
+
+class NotificationsService {
+  // NOTE: Explicitly specifying using keyed properties to ensure
+  // that the types are synced.
+  private readonly repository: NotificationsServiceProps["repository"]
+
+  private readonly siteMember: NotificationsServiceProps["siteMember"]
+
+  constructor({ repository, siteMember }: NotificationsServiceProps) {
+    this.repository = repository
+    this.siteMember = siteMember
+  }
+
+  formatNotifications(notifications: Notification[]) {
+    return notifications.map((notification) => ({
+      message: notification.message,
+      createdAt: notification.createdAt,
+      link: notification.link,
+      isRead: !!notification.firstReadTime,
+      sourceUsername: notification.sourceUsername,
+      type: notification.type,
+    }))
+  }
+
+  async findAll({
+    siteName,
+    userId,
+    findOptions,
+  }: {
+    siteName: string
+    userId: string
+    findOptions?: FindOptions<Notification>
+  }) {
+    return this.repository.findAll({
+      where: {
+        user_id: userId,
+      },
+      order: [
+        ["first_read_time", "DESC NULLS FIRST"],
+        ["priority", "ASC"],
```

Comment:
Sorry one more point, should this priority be descending instead? Meaning that a higher priority should be shown first.

Also, should we also include the notification creation time and sort by descending order? So the sorting order is something like this:
1. Notifications that the user has not seen before will be at the top.
2. Notifications that are more recent are shown first, if the user has not seen it before or multiple notifications have the same first read time.
3. If all else equal, then show notifications with a higher priority first.

Reply:
I've add the time the notification is created to the sorting criteria! But 2 and 3 are swapped - when a user first logs in, they should see in order:

1. Unseen notifications
2. High priority notifications
3. Notifications created more recently

The reason for this is that upon login, any high priority unread notifications should always be at the top of the list, and since reminder notifications are going to only be created once a day, it's very likely for a lower priority notification to be created afterwards.

I've also left the priority direction as is for now - I put it as ascending because I was imagining lower numbers to mean higher priority (so P1 > P2), but I'm open to changing this if it's not intuitive, since the higher priority reminder notifications will only be added later!

Reply:
As discussed offline, both conditions actually need to be fulfilled - 

- For unread notifications, priority should be the determining factor
- For read notifications, they can be sorted by creation time.

This behaviour has been fixed in https://github.com/isomerpages/isomercms-backend/pull/508/commits/7b4e42364dbd6d3c523824d96d01ff574040d280!